src/mechanics/mesh/MeshGmsh.cpp

```diff
@@ -0,0 +1,378 @@
+}
+
+void ProcessSectionASCII(std::istream& file, GmshFileContent& fileContent)
+{
+    std::string line;
+    std::getline(file, line);
+    if (line[0] == '$')
+    {
+        // remove $
+        std::string sectionName = line.substr(1, line.size() - 1);
+        // capitalize
+        std::transform(sectionName.begin(), sectionName.end(), sectionName.begin(), ::toupper);
+
+        // Execute section read function
+        if (sectionName.compare("NODES") == 0)
+            ReadNodesASCII(file, fileContent);
+        else if (sectionName.compare("ELEMENTS") == 0)
+            ReadElementsASCII(file, fileContent);
+        else if (sectionName.compare("PHYSICALNAMES") == 0)
+            ReadPhysicalNamesASCII(file, fileContent);
+        else
+            throw NuTo::Exception(__PRETTY_FUNCTION__, "Unhandled gmsh section type: " + sectionName);
+    }
+}
+
+
+const NuTo::InterpolationSimple& CreateElementInterpolation(NuTo::MeshFem& mesh, int gmshType, int dimension)
+{
+    using namespace NuTo;
+    switch (gmshType)
+    {
+    case 1:
+        return mesh.CreateInterpolation(InterpolationTrussLinear());
+    case 2:
+        return mesh.CreateInterpolation(InterpolationTriangleLinear());
+    case 3:
+        return mesh.CreateInterpolation(InterpolationQuadLinear());
+    default:
+        throw NuTo::Exception(__PRETTY_FUNCTION__, "Unhandled gmsh element type.");
+    }
+}
+
+
+std::string GetPhysicalGroupName(const GmshFileContent& fileContent, int groupId)
+{
+    std::string physicalName{""};
+    for (const GmshPhysicalNames& gmshPhysicalName : fileContent.physicalNames)
+        if (gmshPhysicalName.id == groupId)
+        {
+            physicalName = gmshPhysicalName.name;
+            break;
+        }
+    return physicalName;
+}
+
+
+// Member functions
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+NuTo::MeshGmsh::MeshGmsh(const std::string& fileName)
+{
+    ReadGmshFile(fileName);
+}
+
+const NuTo::Group<NuTo::ElementCollection>& NuTo::MeshGmsh::GetPhysicalGroup(std::string physicalName) const
+{
+    std::transform(physicalName.begin(), physicalName.end(), physicalName.begin(), ::toupper);
+    auto physGroupIt = mNamedPhysicalGroups.find(physicalName);
+    if (physGroupIt == mNamedPhysicalGroups.end())
+        throw Exception(__PRETTY_FUNCTION__, "Couldn't find group with name " + physicalName);
+    return *(physGroupIt->second);
+}
+
+const NuTo::Group<NuTo::ElementCollection>& NuTo::MeshGmsh::GetPhysicalGroup(int physicalGroupId) const
+{
+    auto physGroupIt = mPhysicalGroups.find(physicalGroupId);
+    if (physGroupIt == mPhysicalGroups.end())
+        throw Exception(__PRETTY_FUNCTION__, "Couldn't find group with ID " + std::to_string(physicalGroupId));
+    return physGroupIt->second;
+}
+
+std::vector<NuTo::NodeSimple*> NuTo::MeshGmsh::CreateNodes(const GmshFileContent& fileContent)
```

Comment:
from the gmsh documentation:

> Important note about efficiency. Node and element tags can be "sparse", i.e., do not have to constitute a continuous list of indexes starting at 1. However, using non-continuous tags will lead to performance degradation. For meshes, non-continuous indexing forces Gmsh to use a map instead of a vector to access nodes and elements. The performance hit is on speed. For post-processing datasets, which always use vectors to access data, the performance hit is on memory. A NodeData with two nodes, tagged 1 and 1000000, will allocate a (mostly empty) vector of 1000000 elements. 

The current solution is constant time during lookup and when it is filled. Every other solution is not. The code isn't that much more complicated and when the gmsh class is working, nobody would probably ever look into the code anymore. Apart from that, the solution is part of gmsh too and I think they chose it for a reason ;) 

> no memory allocation erros that you needed to fix - that shows that even for the original authors, the code is not simple enough to avoid the errors

If the code needs to be so simple, that we never make any errors during development, we should stop writing NuTo and focus on `"HelloWorld"` examples. :wink:  - A cookie for the guy who gets an simple hello world example to fail due to memory access violation :trollface:

Reply:
[`unordered_map`](http://en.cppreference.com/w/cpp/container/unordered_map/at#Complexity) has also constant lookup, just saying. 
Its fine, if you do not want to make the change. So, would you mind if I remove the `minNodeId` (and we still use vector)?

Reply:
Why is it so important to remove that? what happens if the first Id is 100000000000 and the second is 100000000001 (which is totally legal for gmsh)? we now have a vector of size 100000000002 instead of 2. 

And to my knowledge `unordered_map` does not have constant lookup. If two elements share the same hash, the "buckets" use linear search algos. But I might be wrong. In general it shouldn't be (that much) slower.

I think we are talking about a small detail, that nobody really cares about and that wouldn't be a performance issue. So if you really want it changed to an `unordered_map`, do it. All I am saying is: I don't see any reason to do it and therefore won't do it myself :wink: